File: 15-3sum/15-3sum.cpp
// Last updated: 7/2/2025, 11:55:42 PM
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        vector<vector<int>> ans;
        sort(nums.begin() , nums.end());
        int len = nums.size(); 

        for(int i=0;i<len-2;i++)
        {
              if( i>0 && nums[i] == nums[i-1]) continue;

              int l = i+1 , r = len-1;
            while(l < r)
            {
                int sum = nums[i] + nums[l] + nums[r];

                if(sum < 0) l++;
                else if(sum > 0) r--;
                else {
                    vector<int> triplet = {nums[i], nums[l], nums[r]};
                    ans.push_back(triplet);

                    while(l+1 < r && nums[l] == nums[l+1]) l++;
                    while(l < r-1 && nums[r] == nums[r-1]) r--;
                    l++;
                    r--;
                }

            }
        }

        return ans;
    }
};

File: 15-3sum/15-3sum.py
# Last updated: 7/2/2025, 11:55:39 PM
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        ans = set()

        n = len(nums)
        l, r = 0, n-1

        for i in range(n-2):
            if(i > 0 and nums[i] == nums[i-1]):
                continue
            l, r = i+1 , n-1
            while l < r:
                totalSum = nums[i] + (nums[l] + nums[r])
                if totalSum == 0:
                    ans.add((nums[i], nums[l], nums[r]))
                    l+=1
                    r-=1
                    while l<r and nums[l] == nums[l-1]:
                        l+=1
                    while l<r and nums[r] == nums[r+1]:
                        r-=1
                elif totalSum > 0:
                    r-=1
                else:
                    l+=1

        return ans
                

        return ans
            
